scipy/signal/ltisys.py

```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        # are chosen to match [1] therefore P_update refers to "Q" from
+        # the QR decomposition and not the permutation matrix usually
+        # named "P".
+        P_update = P_update.T  
+
+        H_start = control_index
+        
+        # find the first occurence of a zero line in R, this will
+        # reveal the rank of H_sub
+        zero_lines_idx = np.where(~R.any(axis=1))[0]
+        if not zero_lines_idx.shape[0]:
+            control_index += np.min(R.shape)
+        else:
+            control_index += zero_lines_idx[0]
+            
+#        It looks like this condition from the paper is useless as the checks
+#        at the end of the loop will break anyway. If a bug is discovered the
+#        first place to look at should be this one.
+#        if control_index == A.shape[0] and H_start != 0:
+#             "if n1 + n2 + ... = n stop" page 175
+#            break 
+            
+#        form P2 = ... page 174 note that for P1 we have H_start=0 so that's
+#        OK even if it is useless   
+        P_update_full = np.zeros(
+            (H_start+P_update.shape[0],H_start+P_update.shape[1])
+            )
+        P_update_full[:H_start,:H_start] = np.eye(H_start)
+        P_update_full[H_start:,H_start:] = P_update
+
+        if P is None:  # P1
+            P = P_update_full
+        else:
+            P = np.dot(P_update_full,P)  # "update P as above page 175"
+
+        # update H and extract the new "H_k_k-1" we will work on
+        H = np.dot(np.dot(P,A),P.T)
+        H_sub = H[control_index:,H_start:control_index]
+            
+        if np.allclose(H_sub,0):  # or H_k_k-1" is a zero matrix         
+            stop = True
+            
+    Bbar = np.dot(P,B)
+    if C is None:
+        Cbar = None
+    else:
+        Cbar = np.dot(C,P.T)
+    return H,Bbar,Cbar,P,control_index
+    
+def controllability_matrix(A,B):
+    r"""
+    Compute the controllability matrix of the linear system ``Xdot=AX+BU``. 
+    If a linear system is controllable it is possible to find a set of inputs
+    U such as any configuration of the state-vector X is reachable in finite
+    time starting from any other configuration.
+    If the rank of the controllability matrix is not equal to the dimension of
+    the state-space the system is not controllable.
+    
+    Parameters
+    ----------
+    A, B : ndarray
+        State-space representation of linear system ``AX + BU``.
+        
+    Returns
+    -------
+    Ctrb_mat : ndarray
+        The controllability matrix
+        
+    Notes
+    -----
+    Even if the rank of the controllability matrix equals the dimension of the
+    state-space the system might be very difficult, if not impossible, to 
+    control, its condition number should also be checked. ONE SHOULD ONLY
+    RELY ON THE RANK OF THIS MATRIX FOR EDUCATIONAL PURPOSES DUE TO THE
+    NUMERICAL INSTABILITY OF THIS COMPUTATION.
+    For serious control design controller_Hessenberg is the way to go.
```

Comment:
undescore has a special meaning for Sphinx. If it's a reference to a routine, single backticks are the way to go.